### PR TITLE
Standardized file admin

### LINKED ISSUE #403: Extend the StandardizedFile admin view to expose button to trigger next step
- [x] Add back [logic to determine whether there is any work being done](https://github.com/datamade/bga-payroll/blob/2178f4ccc8593dd8baa2d4dcbdf5debac0972ca3/data_import/models.py#L148-L173) for a given StandardizedFile
- [x] [Override the StandardizedFile admin detail view](https://docs.djangoproject.com/en/2.2/ref/contrib/admin/#overriding-admin-templates) to display a link or button to advance to the next step, if there is no work being done on the given file

### PR DESCRIPTION
**Pull request by fatima3558**
### Overview
This pull request adds a link to resume the import of a file. 

Closes #403 

### Notes
There is one linting error that I'm not sure how to resolve, with regards to lines 10 and 161 of `data_import/models.py`. This was occasionally giving me errors while I was working; it was telling me that 'str' has no attribute `.get`, which I took to mean it was interpreting `task['kwargs']` in the `processing()` functionality I added back as a string type object. Here's the code for context:

```python
    @property
    def processing(self):
        '''
        Inspect the queue for work related to the StandardizedFile
        at hand. If there is active work, or work on the queue,
        return True. Otherwise, return False.
        '''
        from celery.task.control import inspect

        i = inspect()

        active = i.active()

        for worker, tasks in active.items():
            for task in tasks:
                if task['kwargs'].get('s_file_id') == self.id:
                    return True

        enqueued = i.reserved()

        for worker, tasks in enqueued.items():
            for task in tasks:
                if task['kwargs'].get('s_file_id') == self.id:
                    return True

        return False
```

### CODE REVIEW

**Comment by hancush** [comment]
@fatima3558 Can you resolve these merge conflicts? Taking a look at the rest now.

File: data_import/models.py [hunk 1]
```diff
@@ -144,9 +144,28 @@ def raw_table_name(self):
...
+            for task in tasks:
+                if task['kwargs'].get('s_file_id') == self.id:
+                    return True
+
+        enqueued = i.reserved()
+
+        for worker, tasks in enqueued.items():
+            for task in tasks:
+                if task['kwargs'].get('s_file_id') == self.id:
+                    return True
```
[the remark this answers is not in the document]
Comment (fatima3558) [reply data_import/models.py:167]: gotcha, that makes sense as far as the debugging, will keep that in mind for future bugs.

On doing a quick search, it looks like there are two main ways to parse this string: one by using `json.loads()`, and the other is by using `ast.literal_eval()`.

This project already uses `json.loads()`, so I think it makes sense to use that. In order to implement this, I would parse `tasks['kwargs']` using `json.loads()`, save it to a variable, and use it in lines 159 and 166.

[the remark this answers is not in the document]
Comment (hancush) [reply data_import/models.py:167]: Cool, give it a shot. JSON is very picky about formatting, so if you're running into issues, I'd suggest an eval approach. What's the difference between `ast.literal_eval()` and `eval()`?

File: data_import/models.py [hunk 2]
```diff
@@ -144,9 +146,30 @@ def raw_table_name(self):
...
         '''
+        from celery.task.control import inspect
+
+        i = inspect()
+
+        active = i.active()
+
+        for worker, import_tasks in active.items():
+            for task in import_tasks:
+                kw_args = json.loads(task['kwargs'])
```
Comment (hancush) [inline data_import/models.py:161]: I'm getting this traceback:

```
Traceback:

File "/usr/local/lib/python3.5/site-packages/django/core/handlers/exception.py" in inner
  34.             response = get_response(request)

File "/usr/local/lib/python3.5/site-packages/django/core/handlers/base.py" in _get_response
  115.                 response = self.process_exception_by_middleware(e, request)

File "/usr/local/lib/python3.5/site-packages/django/core/handlers/base.py" in _get_response
  113.                 response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "/usr/local/lib/python3.5/site-packages/django/contrib/admin/options.py" in wrapper
  606.                 return self.admin_site.admin_view(view)(*args, **kwargs)

File "/usr/local/lib/python3.5/site-packages/django/utils/decorators.py" in _wrapped_view
  142.                     response = view_func(request, *args, **kwargs)

File "/usr/local/lib/python3.5/site-packages/django/views/decorators/cache.py" in _wrapped_view_func
  44.         response = view_func(request, *args, **kwargs)

File "/usr/local/lib/python3.5/site-packages/django/contrib/admin/sites.py" in inner
  223.             return view(request, *args, **kwargs)

File "/app/data_import/admin.py" in change_view
  74.         processing = file.processing

File "/app/data_import/models.py" in processing
  161.                 kw_args = json.loads(task['kwargs'])

File "/usr/local/lib/python3.5/json/__init__.py" in loads
  319.         return _default_decoder.decode(s)

File "/usr/local/lib/python3.5/json/decoder.py" in decode
  339.         obj, end = self.raw_decode(s, idx=_w(s, 0).end())

File "/usr/local/lib/python3.5/json/decoder.py" in raw_decode
  355.             obj, end = self.scan_once(s, idx)

Exception Type: JSONDecodeError at /admin/data_import/standardizedfile/1/change/
Exception Value: Expecting property name enclosed in double quotes: line 1 column 2 (char 1)
```

What do you think about an eval approach instead?

Comment (fatima3558) [reply data_import/models.py:161]: Yeah, I can try that. I'm not getting that traceback, though, and I'd like to make sure I can test the eval approach as well. When / how is this exception coming up?

Comment (hancush) [reply data_import/models.py:161]: I uploaded a file, then added all responding agencies, then navigated back in my browser to the standardized file detail page.

**Comment by fatima3558** [comment]
I did see the error message you were getting after I cleared out all the uploads, but even with just one upload, I was still getting this traceback in my terminal:

```bash
bga-payroll-worker | [2020-04-01 21:01:41,572: ERROR/ForkPoolWorker-2] Task data_import.tasks.insert_responding_agency[a62fb8ba-d808-4280-8582-e2a183a2a971] raised unexpected: IntegrityError('duplicate key value violates unique constraint "data_import_standardized_standardizedfile_id_resp_def35295_uniq"\nDETAIL:  Key (standardizedfile_id, respondingagency_id)=(32, 661) already exists.\n',)
bga-payroll-worker | Traceback (most recent call last):
bga-payroll-worker |   File "/usr/local/lib/python3.5/site-packages/django/db/backends/utils.py", line 82, in _execute
bga-payroll-worker |     return self.cursor.execute(sql)
bga-payroll-worker | psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "data_import_standardized_standardizedfile_id_resp_def35295_uniq"
bga-payroll-worker | DETAIL:  Key (standardizedfile_id, respondingagency_id)=(32, 661) already exists.
bga-payroll-worker |
bga-payroll-worker |
bga-payroll-worker | The above exception was the direct cause of the following exception:
bga-payroll-worker |
bga-payroll-worker | Traceback (most recent call last):
bga-payroll-worker |   File "/usr/local/lib/python3.5/site-packages/celery/app/trace.py", line 382, in trace_task
bga-payroll-worker |     R = retval = fun(*args, **kwargs)
bga-payroll-worker |   File "/usr/local/lib/python3.5/site-packages/celery/app/trace.py", line 641, in __protected_call__
bga-payroll-worker |     return self.run(*args, **kwargs)
bga-payroll-worker |   File "/app/data_import/tasks.py", line 129, in insert_responding_agency
bga-payroll-worker |     self.import_utility.insert_responding_agency()
bga-payroll-worker |   File "/app/data_import/utils/import_utility.py", line 75, in insert_responding_agency
bga-payroll-worker |     self._link_responding_agency_with_standardized_file()
bga-payroll-worker |   File "/app/data_import/utils/import_utility.py", line 93, in _link_responding_agency_with_standardized_file
bga-payroll-worker |     cursor.execute(insert)
bga-payroll-worker |   File "/usr/local/lib/python3.5/site-packages/django/db/backends/utils.py", line 99, in execute
bga-payroll-worker |     return super().execute(sql, params)
bga-payroll-worker |   File "/usr/local/lib/python3.5/site-packages/django/db/backends/utils.py", line 67, in execute
bga-payroll-worker |     return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
bga-payroll-worker |   File "/usr/local/lib/python3.5/site-packages/django/db/backends/utils.py", line 76, in _execute_with_wrappers
bga-payroll-worker |     return executor(sql, params, many, context)
bga-payroll-worker |   File "/usr/local/lib/python3.5/site-packages/django/db/backends/utils.py", line 84, in _execute
bga-payroll-worker |     return self.cursor.execute(sql, params)
bga-payroll-worker |   File "/usr/local/lib/python3.5/site-packages/django/db/utils.py", line 89, in __exit__
bga-payroll-worker |     raise dj_exc_value.with_traceback(traceback) from exc_value
bga-payroll-worker |   File "/usr/local/lib/python3.5/site-packages/django/db/backends/utils.py", line 82, in _execute
bga-payroll-worker |     return self.cursor.execute(sql)
bga-payroll-worker | django.db.utils.IntegrityError: duplicate key value violates unique constraint "data_import_standardized_standardizedfile_id_resp_def35295_uniq"
bga-payroll-worker | DETAIL:  Key (standardizedfile_id, respondingagency_id)=(32, 661) already exists.
```

Is it somehow possible deleting the uploads still leaves something of the standardized files behind that's interfering?

**Comment by hancush** [comment]
@fatima3558 Possibly! I'd suggest two things:

- Restart the worker: `docker-compose restart worker`. Sometimes the logs stop streaming, and you're seeing an old error.
- Drop your volume to start totally from scratch: `docker-compose down --volumes && docker-compose up -d`.

**Pull request merged**